hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/AbstractMapQueryMessageTask.java

```diff
@@ -61,66 +63,86 @@ public Permission getRequiredPermission() {
+    private List<Future> createInvocations(Collection<Member> members, Predicate predicate) {
+        List<Future> futures = new ArrayList<Future>(members.size());
         final InternalOperationService operationService = nodeEngine.getOperationService();
         for (Member member : members) {
             Future future = operationService.createInvocationBuilder(SERVICE_NAME,
-                    // todo: this is a performance issue; since always keys + values are downloaded.
-                    new QueryOperation(getDistributedObjectName(), predicate, IterationType.ENTRY),
+                    new QueryOperation(getDistributedObjectName(), predicate, getIterationType()),
                     member.getAddress()).invoke();
             futures.add(future);
         }
+        return futures;
     }
 
-    private void collectResults(Collection<QueryResultRow> result, List<Future> futures, Set<Integer> finishedPartitions)
-            throws InterruptedException, java.util.concurrent.ExecutionException {
-
+    @SuppressWarnings("unchecked")
+    private BitSet collectResults(Collection<QueryResultRow> result, List<Future> futures, int partitionCount)
+            throws InterruptedException, ExecutionException {
+        BitSet finishedPartitions = new BitSet(partitionCount);
         for (Future future : futures) {
             QueryResult queryResult = (QueryResult) future.get();
             if (queryResult != null) {
                 Collection<Integer> partitionIds = queryResult.getPartitionIds();
-                if (partitionIds != null) {
-                    finishedPartitions.addAll(partitionIds);
+                if (partitionIds != null && !hasAtLeastOneBitSet(finishedPartitions, partitionIds)) {
+                    //Collect results only if there is no overlap with already collected partitions.
+                    //If there is an overlap it means there was a partition migration while QueryOperation(s) were
+                    //running. In this case we discard all results from this member and will target the missing
+                    //partition separately later.
+                    BitSetUtils.setBits(finishedPartitions, partitionIds);
                     result.addAll(queryResult.getRows());
                 }
             }
         }
+        return finishedPartitions;
     }
 
-    private boolean hasMissingPartitions(Set<Integer> finishedPartitions, int partitionCount) {
-        return finishedPartitions.size() != partitionCount;
+    private boolean hasMissingPartitions(BitSet finishedPartitions, int partitionCount) {
+        return finishedPartitions.nextClearBit(0) == partitionCount;
     }
 
-    private List<Integer> findMissingPartitions(Set<Integer> finishedPartitions, int partitionCount) {
+    private List<Integer> findMissingPartitions(BitSet finishedPartitions, int partitionCount) {
         List<Integer> missingList = new ArrayList<Integer>();
```

Comment:
Not terribly exciting. but every place where we go through partitions and need to store them, I would go for an array. Arrays are cheap since no wrapper objects are needed + extremely fast is used correctly. And in case of partitions, you always know how big the array needs to be; so there is no need for growing.